### PR TITLE
Bugfix: Fix potential crashes reported from AppStore

### PR DESCRIPTION
**Pull request by wutschel**
## Description

This PR applies 3 confirmed fixes for crashes reported from AppStore (out of bounds access in `didSelectRowAtIndexPath`,  `showAlbumActions` and `layoutAttributesForElementsInRect`).

On top the PR also has two potential fixes which I found after research on www.stackoverflow.com. First is about setting some properties to `nil` in `dealloc` -- I am not sure, if this is relevant anymore. Second, is avoiding a potential recursive call in `textField:shouldChangeCharactersInRange:replacementString:` by using a copy of the object.

## Summary for release notes

Bugfix: Fix potential crashes reported from AppStore

### CODE REVIEW

File: XBMC Remote/DetailViewController.m [hunk 1]
```diff
@@ -4360,9 +4360,24 @@ - (void)showInfo:(NSIndexPath*)indexPath menuItem:(mainMenu*)menuItem item:(NSDi
...
+        LOCALIZED_STR(@"Queue after current"),
+        LOCALIZED_STR(@"Queue"),
+        LOCALIZED_STR(@"Play"),
+        LOCALIZED_STR(@"Play in shuffle mode"),
+        LOCALIZED_STR(@"Album Details"),
+        LOCALIZED_STR(@"Search Wikipedia"),
+    ];
     selected = [NSIndexPath indexPathForRow:0 inSection:0];
-    NSMutableDictionary *item = [NSMutableDictionary dictionaryWithDictionary:[self.sections objectForKey:self.sectionArray[0]][0]];
+    NSMutableDictionary *item = [NSMutableDictionary dictionaryWithDictionary:sectionItem];
```
Comment (kambala-decapitator) [inline XBMC Remote/DetailViewController.m:4380]:
```suggestion
    NSMutableDictionary *item = [sectionItem mutableCopy];
```

**Review by kambala-decapitator — CHANGES_REQUESTED** [review]

File: XBMC Remote/DetailViewController.m [hunk 2]
```diff
@@ -6170,6 +6170,10 @@ - (void)dealloc {
     [self.sections removeAllObjects];
     [channelListUpdateTimer invalidate];
     [[NSNotificationCenter defaultCenter] removeObserver: self];
+    self.searchController.searchResultsUpdater = nil;
+    self.searchController.searchBar.delegate = nil;
+    self.searchController.delegate = nil;
+    self.searchController = nil;
```
[the remark this answers is not in the document]
Comment (wutschel) [reply XBMC Remote/DetailViewController.m:6191]: Done

File: XBMC Remote/XBMCVirtualKeyboard.m [hunk 3]
```diff
@@ -184,7 +184,8 @@ - (BOOL)textField:(UITextField*)theTextField shouldChangeCharactersInRange:(NSRa
         return NO;
     }
     else {
-        NSString *stringToSend = [theTextField.text stringByReplacingCharactersInRange:range withString:string];
+        NSString *stringInTextField = [theTextField.text copy];
+        NSString *stringToSend = [stringInTextField stringByReplacingCharactersInRange:range withString:string];
```
[the remark this answers is not in the document]
Comment (wutschel) [reply XBMC Remote/XBMCVirtualKeyboard.m:188]: Removed

File: XBMC Remote/DetailViewController.m [hunk 4]
```diff
@@ -4360,9 +4360,24 @@ - (void)showInfo:(NSIndexPath*)indexPath menuItem:(mainMenu*)menuItem item:(NSDi
...
 - (void)showAlbumActions:(UITapGestureRecognizer*)tap {
-    NSArray *sheetActions = @[LOCALIZED_STR(@"Queue after current"), LOCALIZED_STR(@"Queue"), LOCALIZED_STR(@"Play"), LOCALIZED_STR(@"Play in shuffle mode"), LOCALIZED_STR(@"Album Details"), LOCALIZED_STR(@"Search Wikipedia")];
+    if (self.sectionArray.count == 0) {
+        return;
+    };
+    id sectionKey = self.sectionArray[0];
+    if ([self.sections[sectionKey] count] == 0) {
+        return;
+    };
+    id sectionItem = self.sections[sectionKey][0];
```
[the remark this answers is not in the document]
Comment (wutschel) [reply XBMC Remote/DetailViewController.m]: Done

File: XBMC Remote/DetailViewController.m (hunk 1 above)
Comment (wutschel) [reply XBMC Remote/DetailViewController.m]: Done

**Pull request merged**